{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import graphviz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a function which can return consective k-mers of a string\n",
    "def K_mers(str:str, k:int)->list:\n",
    "    \"\"\"\n",
    "    str: a string from which k-mers have to be constructed\n",
    "    k:   length of each k-mer\n",
    "    \"\"\"\n",
    "    k_mers = []\n",
    "    for i in range(len(str) - k + 1):\n",
    "        k_mers.append(str[i: i + k])\n",
    "    \n",
    "    return k_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prefix(string:str)->str:\n",
    "    \"\"\"it returns prefix of the string leaving only the last char by default e.g. string[:-1]\n",
    "    string: a string for which prefix is required\n",
    "    returns:\n",
    "        a prefix of string \n",
    "    \"\"\"\n",
    "    return string[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sufix(string:str)->str:\n",
    "    \"\"\"it returns sufix of the string leaving only the first char by default e.g. string[1:]\n",
    "    string: a string for which sufix is required\n",
    "    returns:\n",
    "        a sufix of string \n",
    "    \"\"\"\n",
    "    return string[1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(file_name: str) -> str:\n",
    "    \"\"\"\n",
    "    file_name: file from which genome is to be read\n",
    "    returns:\n",
    "        genome as a string\n",
    "    \"\"\"\n",
    "    # loade the data from .fasta file\n",
    "    file = open(file_name, \"r\")\n",
    "    # reading the whole file\n",
    "    raw_data = file.read()\n",
    "\n",
    "    # doing data clean-up if needed\n",
    "    # if we find \">\" at the start we need to remove this line\n",
    "    # this first line only for mata-data\n",
    "    if raw_data[0] == \">\":\n",
    "        # finding the end of first line by raw_data.find(\"\\n\")\n",
    "        # returns the index of first match\n",
    "        # keep all the string except first line\n",
    "        data = raw_data[raw_data.find(\"\\n\"):]\n",
    "\n",
    "    # removing all the endline parameters from the string\n",
    "    # this will give a long complete string of given refrence genome\n",
    "    data = data.replace(\"\\n\", \"\")\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Node:\n",
    "    \"\"\"\n",
    "    It save the text value of the node along with out and in going\n",
    "    edges in it.\n",
    "    \"\"\"\n",
    "    def __init__(self, value:str) -> None:\n",
    "        \"\"\"\n",
    "        value: any value to store in a Node\n",
    "        \"\"\"\n",
    "        # text value stored in the node\n",
    "        self.Value = value\n",
    "        # number of all the incomming edges to this node\n",
    "        self.inCommingEdges = 0\n",
    "        # number of all the out going edges from this node\n",
    "        self.outGoingEdges = 0\n",
    "    \n",
    "    def isBalanced(self)->bool:\n",
    "        \"\"\"This function returns True if Node is Balance (if number \n",
    "        of in comming edges is equla to\n",
    "        number of out going edges then node is balanced)\n",
    "        otherwise it retuns False.\"\"\"\n",
    "        if self.inCommingEdges == self.outGoingEdges:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    \n",
    "    # over-riding default hash function so Nodes can be used as key in\n",
    "    # dict in python\n",
    "    def __hash__(self) -> int:\n",
    "        return hash(self.Value)\n",
    "    \n",
    "    # over-riding default print() function for Node\n",
    "    def __str__(self) -> str:\n",
    "        return self.Value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DeBruijnGraph:\n",
    "    \"\"\"This Contain De Bruijn Graph implementation with all the helper\n",
    "    functions\"\"\"\n",
    "    def __init__(self) -> None:\n",
    "        # inilizing all the varibales with defalut values\n",
    "        \n",
    "        # this dict will store key value pairs as shown\n",
    "        # {\n",
    "        #     \"string_1\":[Node(), Node()],\n",
    "        #     \"string_1\":[Node(],\n",
    "        # }\n",
    "        self.Graph = {}\n",
    "        \n",
    "        # a list of all the nodes in the graph\n",
    "        # {\n",
    "        #   \"string_1\":Node(\"string_1\")\n",
    "        # }\n",
    "        self.nodes = {}\n",
    "\n",
    "        # to keep track of all the balamced nodes in the graph\n",
    "        self.totalBalanceNodes = 0\n",
    "\n",
    "        # total nodes in the graph\n",
    "        self.totalNodes = 0\n",
    "    \n",
    "    def addNode(self, value:str)->None:\n",
    "        \"\"\"it adds a new node in the graph and inilize its values by \n",
    "        empty list, if node exist it do nothing\n",
    "        value: a string value which will be stored into the new Node\n",
    "        returns: it returns nothing\"\"\"\n",
    "\n",
    "        # if node not exist already\n",
    "        if value not in self.Graph.keys():\n",
    "            # add new node\n",
    "            self.Graph[value] = []\n",
    "        if value not in self.nodes.keys():\n",
    "            # add new node to the text graph also\n",
    "            self.nodes[value] = Node(value)\n",
    "    \n",
    "    def addEdge(self, from_node:str, to_node:str)->None:\n",
    "        \"\"\"it will add a new edge from a node (from_node) to a node (to_node)\n",
    "        from_node: a string value of the starting node of the edge\n",
    "        to_node: a string value of the ending node of the edge\n",
    "        \"\"\"\n",
    "        # updating Node list\n",
    "        if from_node in self.nodes.keys():\n",
    "            self.nodes[from_node].outGoingEdges += 1\n",
    "        else:\n",
    "            temp_node = Node(from_node)\n",
    "            temp_node.outGoingEdges += 1\n",
    "            self.nodes[from_node] = temp_node\n",
    "        \n",
    "        if to_node in self.nodes.keys():\n",
    "            self.nodes[to_node].inCommingEdges += 1\n",
    "        else:\n",
    "            temp_node = Node(to_node)\n",
    "            temp_node.inCommingEdges += 1\n",
    "            self.nodes[to_node] = temp_node\n",
    "\n",
    "        # if node already exits\n",
    "        if from_node in self.Graph.keys():\n",
    "            # appending a new node to already present list\n",
    "            self.Graph[from_node].append(Node(to_node))\n",
    "        else:\n",
    "            # if not present\n",
    "            # inilizing with a list which conain one node\n",
    "            self.Graph[from_node] = [Node(to_node)]\n",
    "    \n",
    "    def buildFromFile(self, filename:str, k:int = 12):\n",
    "        \"\"\"\"it takes .fasta file name from which to read and build the graph\n",
    "        filename: file from which data has to read with extention\n",
    "        k: size of each k-mer to be taken by default it is 12\n",
    "        returns: it returns nothing\"\"\"\n",
    "        # reading the file and cleaning the data\n",
    "        data = readGenome(filename)\n",
    "        # choping the data into k-mers of specified lenght\n",
    "        k_mers = K_mers(data, k)\n",
    "        # sorting the k-mers\n",
    "        k_mers.sort()\n",
    "        # adding each k-mer to the graph\n",
    "        for each in k_mers:\n",
    "            self.addEdge(prefix(each), sufix(each))\n",
    "        \n",
    "        # updating total number of nodes\n",
    "        self.totalNodes = len(list(self.nodes.keys()))\n",
    "\n",
    "        # updating total balanced nodes in the graph\n",
    "        t_list = []\n",
    "        for eachNode in self.nodes.keys():\n",
    "            if self.nodes[eachNode].isBalanced():\n",
    "                self.totalBalanceNodes += 1\n",
    "            else:\n",
    "                t_list.append(eachNode)\n",
    "        print(\"Un Balanced Nodes\")\n",
    "        print(t_list)\n",
    "\n",
    "    def eulerCycle(self):\n",
    "        \"\"\"\"it returns a Eulerian Path from the graph\"\"\"\n",
    "        pass\n",
    "\n",
    "    def isEulerian(self)->bool:\n",
    "        \"\"\"it returns true if graph is Eulerian graph\"\"\"\n",
    "        pass\n",
    "\n",
    "    # this function render graph to a pdf file for view\n",
    "    def renderGraph(self, filename:str):\n",
    "        g = graphviz.Digraph(format='pdf')\n",
    "        for eachNode in self.Graph.keys():\n",
    "            for eachToNode in self.Graph[eachNode]:\n",
    "                g.edge(eachNode, eachToNode.Value, label=\"\")\n",
    "        g.render(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = DeBruijnGraph()\n",
    "g.buildFromFile(\"../../Data/temp_data.fasta\", k = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g.renderGraph(\"./graph\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "3fab1b6aec8adb0f66c1d94bfaed860b0c54ac70f3c4b7785c313af9c15b40f3"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('fyp_env': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
